patches/server/1036-Expose-native-map-methods-in-PDC.patch

```diff
@@ -0,0 +1,250 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Xander <xander.dekeijzer@student.hu.nl>
+Co-authored-by: TheRealRyGuy <ryan@ryguy.me>
+Date: Tue, 3 Oct 2023 07:29:00 +0200
+Subject: [PATCH] Expose native map methods in PDC
+
+
+diff --git a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
+index 91cb4a28afa5d83e6de10dab834ed63e2eb3b76f..4a3e137866611357137edb36cf76c67885957e91 100644
+--- a/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
++++ b/src/main/java/org/bukkit/craftbukkit/persistence/CraftPersistentDataContainer.java
+@@ -7,6 +7,10 @@ import java.util.Map;
+ import java.util.Map.Entry;
+ import java.util.Objects;
+ import java.util.Set;
++import java.util.function.BiFunction;
++import java.util.function.Function;
++import java.util.function.Supplier;
++import net.kyori.adventure.key.Key;
+ import net.minecraft.nbt.CompoundTag;
+ import net.minecraft.nbt.Tag;
+ import org.bukkit.NamespacedKey;
+@@ -14,6 +18,8 @@ import org.bukkit.craftbukkit.util.CraftNBTTagConfigSerializer;
+ import org.bukkit.persistence.PersistentDataAdapterContext;
+ import org.bukkit.persistence.PersistentDataContainer;
+ import org.bukkit.persistence.PersistentDataType;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
+ 
+ public class CraftPersistentDataContainer implements PersistentDataContainer {
+ 
+@@ -37,16 +43,27 @@ public class CraftPersistentDataContainer implements PersistentDataContainer {
+         Preconditions.checkArgument(key != null, "The NamespacedKey key cannot be null");
+         Preconditions.checkArgument(type != null, "The provided type cannot be null");
+         Preconditions.checkArgument(value != null, "The provided value cannot be null");
++        // Paper start
+ 
+-        this.customDataTags.put(key.toString(), this.registry.wrap(type.getPrimitiveType(), type.toPrimitive(value, adapterContext)));
++        set(key.toString(), type, value);
++    }
++
++    private <T, Z> void set(String key, PersistentDataType<T, Z> type, Z value) {
```

Comment:
Instead of taking a String, these methods can actually just take an adventure `Key`. That way the entire method body of `get`, `has`, `remove`, etc. can be moved into the private methods, and the public implementations can delegate to those private methods. It simplifies the diff dramatically. Make sure to change the toString() methods to asString() for the adventure Key tho.